### DISCUSSION TITLE
Reading a TDMS file on Apache Athena

### DISCUSSION THREAD

**Discussion by wz2b** [adamreeve/npTDMS#305] [Q&A]
Hi, I am trying to use nptdms on apache athena (which is really pyspark).  I was able to install nptdms on athena with no real problems.  Now, as I try to parallelize/iterate through files, I'm not sure what to do next.  In Athena, you load binary files using their version of a DataFrame, where one of the rows of the dataframe is _content_ - basically it's a bytearray.

How do I load that into nptdms?  I don't see an obvious way to just feed it the bytes.

--C

**Comment by adamreeve** [comment]
Hi @wz2b, npTDMS works with any file like object. You'll probably want to create an `io.BytesIO` instance from your bytearray to read it into npTDMS. See https://docs.python.org/3/library/io.html#io.BytesIO

**Comment by wz2b** [comment]
Do you think there's a way to feed an s3://bucket/whatever.tdms stream
right into it?

**Reply by adamreeve** [reply]
What type of Python object is your S3 stream?

**Comment by wz2b** [comment]
From the AWS client s3 library it would be a StreamingBody:

https://botocore.amazonaws.com/v1/documentation/api/latest/reference/response.html#botocore.response.StreamingBody

which does have a method named read()  ... so it seems like that might work

**Reply by adamreeve** [reply]
Ah actually I don't think this will work as npTDMS needs to be able to seek back to arbitrary positions in the file, so won't work when streaming (and the StreamingBody class doesn't implement the `seek` method). So you will need to read the whole file into a bytearray or bytes first.
